apps/api/src/persistence/repositories/repository-for-entity.ts

```diff
@@ -8,46 +9,60 @@ import { isNotFound, NotFound } from '../../lib/types/not-found';
 import { ResultOrError } from '../../types/ResultOrError';
 import { ArangoDatabaseForCollection } from '../database/arango-database-for-collection';
 import { DatabaseProvider } from '../database/database.provider';
-import { ArangoResourceCollectionID } from '../database/types/ArangoCollectionId';
+import { ArangoCollectionID } from '../database/types/ArangoCollectionId';
 import mapDatabaseDTOToEntityDTO from '../database/utilities/mapDatabaseDTOToEntityDTO';
-import mapEntityDTOToDatabaseDTO from '../database/utilities/mapEntityDTOToDatabaseDTO';
+import mapEntityDTOToDatabaseDTO, {
+    DatabaseDTO,
+} from '../database/utilities/mapEntityDTOToDatabaseDTO';
 
 /**
  * TODO We need to add error handling. It is especially important that if
  * the `instance factory` fails to build an instance because the dto violates
  * the model invariants that an easy to understand error is returned.
+ *
+ * TODO Use a mixin for cloneable behaviour.
  */
-export class RepositoryForEntity<TEntity extends Resource>
+export class RepositoryForEntity<TEntity extends HasEntityID & BaseDomainModel>
     implements IRepositoryForEntity<TEntity>
 {
     #arangoDatabaseForEntitysCollection: ArangoDatabaseForCollection<TEntity>;
 
     // Typically just uses the model constructor
     #instanceFactory: InstanceFactory<TEntity>;
 
+    #mapDocumentToEntityDTO;
+
+    #mapEntityDTOToDocument;
+
     constructor(
         arangoDatabaseProvider: DatabaseProvider,
-        collectionName: ArangoResourceCollectionID,
-        instanceFactory: InstanceFactory<TEntity>
+        collectionName: ArangoCollectionID,
+        instanceFactory: InstanceFactory<TEntity>,
+        documentToEntity = mapDatabaseDTOToEntityDTO,
```

Comment:
This whole approach is a bit confusing. Either
- clean it up
- find another way to do it (what other approaches did I consider at the time?)

Reply:
cleaned up